library/std/src/sys/unix/kernel_copy.rs

```diff
@@ -114,11 +114,11 @@ impl FdMeta {
         }
     }
 
-    fn copy_file_range_candidate(&self) -> bool {
+    fn copy_file_range_candidate(&self, is_read: bool) -> bool {
         match self {
             // copy_file_range will fail on empty procfs files. `read` can determine whether EOF has been reached
             // without extra cost and skip the write, thus there is no benefit in attempting copy_file_range
-            FdMeta::Metadata(meta) if meta.is_file() && meta.len() > 0 => true,
+            FdMeta::Metadata(meta) if meta.is_file() && (!is_read || meta.len() > 0) => true,
             FdMeta::NoneObtained => true,
```

Comment:
Now seeing the code I think the is_read is a bit too subtle. `NoneObtained` kind of, sort of used to imply `!is_read` but this was changed recently. Now `NoneObtained` should only happen when stat fails so in that case we probably shouldn't attempt copy_file_range either.

So it probably makes sense to split this into two methods that cover the source/sink case separately.

Reply:
I've made another attempt at doing this DRY-ly using a more descriptive enum (w/ `PartialEq`) just for this purpose - let me know what you think, I don't love the naming here (`FdHandle`) but it seemed apt enough.

I agree that the previous code was a bit obtuse, in fact I had the boolean logic reversed in my first attempt because it's not abundantly clear. This iteration with the enum is more explicit and easier to reason about at least.